### PR TITLE
Create 0606-construct-string-from-binary-tree.js

### PR DESCRIPTION
**Pull request by aadil42**
Solved construct-string-from-binary-tree in JS.

File(s) Added: 0606-construct-string-from-binary-tree.js
Language(s) Used: JavaScript
Submission URL: https://leetcode.com/problems/construct-string-from-binary-tree/submissions/1354480720/

### CODE REVIEW

File: javascript/0606-construct-string-from-binary-tree.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ *     this.right = (right===undefined ? null : right)
+ * }
+ */
+/**
+ * @param {TreeNode} root
+ * @return {string}
+ */
+var tree2str = function(root) {
+
+    let str = "";
```
Comment (aakhtar3) [inline javascript/0606-construct-string-from-binary-tree.js:18]: Should avoid string due to performance issues

```js

/**
 * DFS - Post Order
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/construct-string-from-binary-tree/
 * @param {TreeNode} root
 * @return {string}
*/
var tree2str = (root) => {
    const isBaseCase = (root === null);
    if (isBaseCase) return '';
    
    const isLeaf = ((root.left === null) && (root.right === null));
    if (isLeaf) return `${root.val}`;
    
    const left = tree2str(root.left);
    const right = tree2str(root.right);
    
    const isMissingRight = (root.right === null);
    if (isMissingRight) return `${root.val}(${left})`;
    
    return `${root.val}(${left})(${right})`;
};

/**
 * DFS - Iteration
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/construct-string-from-binary-tree/
 * @param {TreeNode} root
 * @return {string}
*/
var tree2str = (root, stack = [ root ]) => {
    const isBaseCase = (root === null);
    if (isBaseCase) return '';
    
    return dfsIteration(stack);
}

const dfsIteration = (stack, buffer = [], visited = new Set()) => {
    while (stack.length !== 0) {
        const root = stack.at(-1);

        if (visited.has(root)) {
            stack.pop();
            buffer.push(')');

            continue;
        }
        
        if (root.right) stack.push(root.right);
        if (root.left) stack.push(root.left);
        
        visited.add(root);
        buffer.push(`(${root.val}`);
        
        const isMissingLeft = ((root.left === null) && (root.right !== null))
        if (isMissingLeft) buffer.push('()');
    }
    
    return buffer
        .join('')
        .slice(1, -1);
}
```

Comment (aadil42) [reply javascript/0606-construct-string-from-binary-tree.js:18]: How about this approach?

Here's the submission link of this code:
https://leetcode.com/problems/construct-string-from-binary-tree/submissions/1388679375/

```javascript
/**
 * InOrder Traversal 
 * Time O(n) | Space O(n)
 * https://leetcode.com/problems/construct-string-from-binary-tree/
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string}
 */
var tree2str = function(root) {
    return dfs(root, []).join("");
};

const dfs = (node, strArr) => {
    if(!node) return;

    strArr.push(node.val);

    if(node.right  || node.left) strArr.push("(");
    dfs(node.left, strArr);
    if(node.right || node.left) strArr.push(")");

    // right tree
    if(node.right) strArr.push("(");
    dfs(node.right, strArr);
    if(node.right) strArr.push(")");

    return strArr;
}
```

Comment (aadil42) [reply javascript/0606-construct-string-from-binary-tree.js:18]: @aakhtar3 (tagging to draw attention to the proposed code)

Comment (Ykhan799) [reply javascript/0606-construct-string-from-binary-tree.js]: @aadil42 For line 25, there's an extra space between node.right and the OR operator. Delete the extra space. Other than that, the rest of the solution looks good 👍

Comment (aadil42) [reply javascript/0606-construct-string-from-binary-tree.js]: @Ykhan799, it's done.

**Pull request merged**